File: extension/src/quicklinks.js
// Quick-links offline op queue — OWNED BY THE SERVICE WORKER (§6/§10).
//
// The startpage asks the SW to enqueue an op (runtime.sendMessage
// `enqueue_quicklink_op`). The SW: (1) appends it to the durable queue in
// chrome.storage.local (offline can last DAYS — storage.session would die with the
// browser, §10); (2) OPTIMISTICALLY edits the cached quick_links so the link shows
// immediately even before any flush (§10 "Постановка в очередь сразу правит кэш");
// (3) best-effort flushes the whole queue to POST /api/quick_links/ops with a
// STABLE Idempotency-Key, so a retried flush is a no-op server-side (§10).
//
// Dependency-injected via `env` so it runs under the SW (real chrome/fetch) and
// under vitest (a fake env). Pure helpers are exported for unit tests.

export const STATE_CACHE_KEY = "stateCache"; // { state: StateResponse, cached_at }
export const QUEUE_KEY = "quickLinkQueue"; // { key, ops: [...] }

// ws://→http://, wss://→https://; trailing slashes trimmed (same mapping as popup).
export function httpBaseFromServiceUrl(serviceUrl) {
  const base = String(serviceUrl || "").replace(/\/+$/, "");
  if (base.startsWith("wss://")) return "https://" + base.slice("wss://".length);
  if (base.startsWith("ws://")) return "http://" + base.slice("ws://".length);
  return base;
}

// Apply ONE op to a quick-links array (NEW array). Mirrors the server + startpage
// semantics: add appends (url upsert on title), remove by id/url, reorder by id list.
export function applyOpToQuickLinks(list, op) {
  const links = (list || []).map((l) => ({ ...l }));
  if (!op || typeof op !== "object") return links;
  if (op.op === "add") {
    if (!op.url) return links;
    const existing = links.find((l) => l.url === op.url);
    if (existing) {
      if (op.title !== undefined) existing.title = op.title;
      return links;
    }
    const position = links.reduce((m, l) => Math.max(m, l.position ?? -1), -1) + 1;
    links.push({ id: op.id ?? null, url: op.url, title: op.title ?? null, position, pending: true });
    return links;
  }
  if (op.op === "remove") {
    return links.filter((l) => (op.id != null ? l.id !== op.id : l.url !== op.url));
  }
  if (op.op === "reorder") {
    if (!Array.isArray(op.order)) return links;
    const byId = new Map(links.map((l) => [l.id, l]));
    const ordered = [];
    op.order.forEach((id, i) => {
      const l = byId.get(id);
      if (l) {
        l.position = i;
        ordered.push(l);
        byId.delete(id);
      }
    });
    for (const l of byId.values()) ordered.push(l);
    return ordered;
  }
  return links;
}

async function readQueue(env) {
  const got = await env.storageLocalGet(QUEUE_KEY);
  const q = got && got[QUEUE_KEY];
  if (q && Array.isArray(q.ops)) return q;
  return { key: null, ops: [] };
}

// Enqueue one op: durable queue + optimistic cache edit. Returns the queue.
export async function enqueueOp(env, op) {
  const q = await readQueue(env);
  // A STABLE Idempotency-Key per accumulation window: generated when the first op
  // lands and kept until a successful flush clears the queue, so retries of the
  // SAME batch dedupe server-side (§10).
  if (!q.key) q.key = env.randomUUID();
  q.ops.push(op);
  await env.storageLocalSet({ [QUEUE_KEY]: q });

  // Optimistic cache edit (§10): the shown quick_links change immediately.
  const cacheGot = await env.storageLocalGet(STATE_CACHE_KEY);
  const wrap = cacheGot && cacheGot[STATE_CACHE_KEY];
  if (wrap && wrap.state) {
    wrap.state.quick_links = applyOpToQuickLinks(wrap.state.quick_links, op);
    await env.storageLocalSet({ [STATE_CACHE_KEY]: wrap });
  }
  return q;
}

// Flush the whole queue to POST /api/quick_links/ops with the stable Idempotency-Key.
// On success: reconcile the cache with the server's authoritative quick_links and
// clear the queue. On any failure: leave the queue for a later retry. Never throws.
export async function flushQueue(env) {
  const q = await readQueue(env);
  if (q.ops.length === 0) return { flushed: false };
  let config;
  try {
    config = await env.getInstanceConfig();
  } catch {
    return { flushed: false, error: "config" };
  }
  const base = httpBaseFromServiceUrl(config.serviceUrl);
  try {
    const resp = await env.fetchFn(base + "/api/quick_links/ops", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer " + config.token,
        "Idempotency-Key": q.key,
      },
      body: JSON.stringify(q.ops),
    });
    if (!resp.ok) return { flushed: false, error: "http_" + resp.status };
    const body = await resp.json();
    // Reconcile the cache with the server's authoritative list.
    const cacheGot = await env.storageLocalGet(STATE_CACHE_KEY);
    const wrap = (cacheGot && cacheGot[STATE_CACHE_KEY]) || { state: {}, cached_at: env.now() };
    wrap.state = wrap.state || {};
    if (body && Array.isArray(body.quick_links)) {
      wrap.state.quick_links = body.quick_links;
      await env.storageLocalSet({ [STATE_CACHE_KEY]: wrap });
    }
    // Clear the queue (and its key) so the next window gets a fresh Idempotency-Key.
    await env.storageLocalSet({ [QUEUE_KEY]: { key: null, ops: [] } });
    return { flushed: true, quick_links: body && body.quick_links };
  } catch {
    return { flushed: false, error: "network" };
  }
}

File: src/api/quick_links.py
"""``POST /api/quick_links/ops`` (§10).

The startpage owns an OFFLINE op queue in ``chrome.storage.local`` (offline can
last days, §10) and flushes it here as an ARRAY of ops with an ``Idempotency-Key``
header. Conflict resolution is last-write-wins by receive order; ``position`` is
recomputed SERVER-side (append); ``reorder`` is the explicit reposition op.

The whole batch — the idempotency check and every op — applies in ONE
``Database.write`` transaction (Фаза 2 contract: parameterized SQL, no ``await``
inside), so a retried flush can never half-apply. The response is the resulting
quick-links snapshot so the page can reconcile its optimistic cache.
"""

from __future__ import annotations

import time

from starlette.exceptions import HTTPException
from starlette.requests import Request
from starlette.responses import JSONResponse

from src.api.guards import require_api_caller, require_not_paused, require_operational
from src.db import quick_links as ql


def _now_ms() -> int:
    return int(time.time() * 1000)


async def quick_links_ops(request: Request) -> JSONResponse:
    await require_api_caller(request)
    require_operational(request)
    await require_not_paused(request)  # pause silences the offline op flush too (§7)

    try:
        ops = await request.json()
    except Exception:
        raise HTTPException(status_code=400, detail="request body must be JSON")
    if not isinstance(ops, list):
        raise HTTPException(
            status_code=400, detail="request body must be a JSON array of ops"
        )

    # Idempotency-Key dedupes a retried flush of the SAME batch (§10). Optional: a
    # missing key just means "always apply" (a client that does not retry).
    key = request.headers.get("idempotency-key") or None

    quick_links = await request.app.state.db.write(
        lambda c: ql.apply_ops_with_key(c, ops, key, _now_ms())
    )
    return JSONResponse(
        {
            "ok": True,
            "quick_links": [
                {"id": r["id"], "url": r["url"], "title": r["title"], "position": r["position"]}
                for r in quick_links
            ],
        }
    )
